File: src/gui/components/timing_calibration.py
"""
Timing Calibration Dialog for NC-KTV
Helps users automatically detect and fix timing drift
"""

from PyQt6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QPushButton,
    QLabel, QSpinBox, QDoubleSpinBox, QGroupBox, QMessageBox
)
from PyQt6.QtCore import Qt
import logging

logger = logging.getLogger(__name__)


class TimingCalibrationDialog(QDialog):
    """Dialog for automatic timing calibration"""
    
    def __init__(self, project, parent=None):
        super().__init__(parent)
        self.project = project
        self.setWindowTitle("Timing Calibration")
        self.setMinimumWidth(500)
        
        self._init_ui()
        self._check_sample_rates()
    
    def _init_ui(self):
        """Initialize UI"""
        layout = QVBoxLayout(self)
        
        # Title
        title = QLabel("🎯 Timing Synchronization")
        title.setStyleSheet("font-size: 16pt; font-weight: bold;")
        layout.addWidget(title)
        
        # Info
        info = QLabel(
            "This tool helps eliminate timing drift between vocal separation, "
            "transcription, and playback."
        )
        info.setWordWrap(True)
        layout.addWidget(info)
        
        layout.addSpacing(15)
        
        # Sample Rate Check
        sample_group = QGroupBox("1. Sample Rate Validation")
        sample_layout = QVBoxLayout()
        
        self.sample_status_label = QLabel("Checking...")
        sample_layout.addWidget(self.sample_status_label)
        
        self.sample_detail_label = QLabel("")
        self.sample_detail_label.setWordWrap(True)
        self.sample_detail_label.setStyleSheet("color: #666; font-size: 9pt;")
        sample_layout.addWidget(self.sample_detail_label)
        
        sample_group.setLayout(sample_layout)
        layout.addWidget(sample_group)
        
        # Global Offset
        offset_group = QGroupBox("2. Global Timing Offset")
        offset_layout = QVBoxLayout()
        
        offset_info = QLabel("Adjust if lyrics consistently lag or lead vocals:")
        offset_layout.addWidget(offset_info)
        
        offset_controls = QHBoxLayout()
        offset_controls.addWidget(QLabel("Offset (seconds):"))
        
        self.offset_spinbox = QDoubleSpinBox()
        self.offset_spinbox.setRange(-5.0, 5.0)
        self.offset_spinbox.setSingleStep(0.01)
        self.offset_spinbox.setDecimals(3)
        self.offset_spinbox.setValue(
            self.project.timing_metadata.get('global_offset', 0.0)
        )
        offset_controls.addWidget(self.offset_spinbox)
        
        offset_controls.addWidget(QLabel("(negative = earlier, positive = later)"))
        offset_controls.addStretch()
        offset_layout.addLayout(offset_controls)
        
        # Auto-calibrate button
        self.auto_btn = QPushButton("🎵 Auto-Calibrate")
        self.auto_btn.setToolTip(
            "Play first lyrics line and press SPACE when you hear it. "
            "The offset will be calculated automatically."
        )
        self.auto_btn.clicked.connect(self._start_auto_calibrate)
        offset_layout.addWidget(self.auto_btn)
        
        offset_group.setLayout(offset_layout)
        layout.addWidget(offset_group)
        
        # Transcription Source
        transcription_group = QGroupBox("3. Transcription Source")
        transcription_layout = QVBoxLayout()
        
        transcription_info = QLabel(
            "✅ RECOMMENDED: Use 'Original Audio' for transcription\n"
            "This eliminates timing offset from vocal separation."
        )
        transcription_info.setStyleSheet("color: #0a0; font-weight: bold;")
        transcription_layout.addWidget(transcription_info)
        
        current_source = self.project.timing_metadata.get('transcription_source', 'unknown')
        source_label = QLabel(f"Current setting: {current_source}")
        transcription_layout.addWidget(source_label)
        
        transcription_group.setLayout(transcription_layout)
        layout.addWidget(transcription_group)
        
        layout.addSpacing(15)
        
        # Buttons
        btn_layout = QHBoxLayout()
        btn_layout.addStretch()
        
        btn_apply = QPushButton("Apply")
        btn_apply.clicked.connect(self.accept)
        btn_layout.addWidget(btn_apply)
        
        btn_cancel = QPushButton("Cancel")
        btn_cancel.clicked.connect(self.reject)
        btn_layout.addWidget(btn_cancel)
        
        layout.addLayout(btn_layout)
    
    def _check_sample_rates(self):
        """Check sample rates of all audio files"""
        try:
            results = self.project.validate_sample_rates_sync()
            
            if results['valid']:
                sr = results['recommended_rate']
                self.sample_status_label.setText(f"✅ All files match: {sr} Hz")
                self.sample_status_label.setStyleSheet("color: green; font-weight: bold;")
                
                files = results['sample_rates']
                details = "Files checked:\n" + "\n".join(
                    f"  • {name}: {rate} Hz" 
                    for name, rate in files.items() 
                    if rate
                )
                self.sample_detail_label.setText(details)
            else:
                mismatches = results['mismatches']
                self.sample_status_label.setText(f"⚠️ Sample rate mismatch detected!")
                self.sample_status_label.setStyleSheet("color: orange; font-weight: bold;")
                
                files = results['sample_rates']
                details = (
                    f"Found different rates: {mismatches}\n\n"
                    "Files:\n" + "\n".join(
                        f"  • {name}: {rate} Hz" 
                        for name, rate in files.items() 
                        if rate
                    ) + f"\n\nRecommended: Resample all to {results['recommended_rate']} Hz"
                )
                self.sample_detail_label.setText(details)
                
        except Exception as e:
            logger.error(f"Sample rate check failed: {e}")
            self.sample_status_label.setText("⚠️ Unable to check sample rates")
            self.sample_detail_label.setText(str(e))
    
    def _start_auto_calibrate(self):
        """Start auto-calibration process"""
        QMessageBox.information(
            self,
            "Auto-Calibration",
            "Auto-calibration feature coming soon!\n\n"
            "For now, manually adjust the offset:\n"
            "• If lyrics appear LATE: Use negative offset (e.g., -0.5)\n"
            "• If lyrics appear EARLY: Use positive offset (e.g., +0.5)\n\n"
            "Start with small adjustments (±0.1s) and fine-tune."
        )
    
    def get_global_offset(self) -> float:
        """Get the configured global offset"""
        return self.offset_spinbox.value()
    
    def apply_settings(self):
        """Apply timing settings to project"""
        offset = self.get_global_offset()
        self.project.timing_metadata['global_offset'] = offset
        self.project.audio_clock.set_master_offset(offset)
        logger.info(f"Applied global offset: {offset:.3f}s")

File: cpp/src/gui/components/timing_calibration.cpp
/*
 * NC-KTV GUI — Timing Calibration Implementation
 */

#include "timing_calibration.h"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>
#include <QGroupBox>

namespace ncktv {

TimingCalibration::TimingCalibration(QWidget* parent) : QWidget(parent) {
    setupUi();
    applyTheme();
}

void TimingCalibration::setupUi() {
    auto* mainLayout = new QVBoxLayout(this);
    mainLayout->setSpacing(12);
    mainLayout->setContentsMargins(12, 12, 12, 12);

    auto* titleLabel = new QLabel("🎯 Timing Calibration", this);
    titleLabel->setStyleSheet("font-size:14px; font-weight:bold; color:#ccc; border:none;");
    mainLayout->addWidget(titleLabel);

    // ── Master Offset ────────────────────────────────────────────────────
    auto* offsetGroup = new QGroupBox("Global Timing Offset", this);
    auto* offsetLayout = new QVBoxLayout(offsetGroup);
    offsetLayout->setContentsMargins(12, 16, 12, 10);

    m_offsetLabel = new QLabel("0.000 s", this);
    m_offsetLabel->setAlignment(Qt::AlignCenter);
    m_offsetLabel->setStyleSheet("font-size:24px; font-weight:bold; color:#22d3ee; border:none;");
    offsetLayout->addWidget(m_offsetLabel);

    m_offsetSlider = new QSlider(Qt::Horizontal, this);
    m_offsetSlider->setRange(-5000, 5000); // ±5.000 seconds in milliseconds
    m_offsetSlider->setValue(0);
    m_offsetSlider->setTickInterval(1000);
    m_offsetSlider->setTickPosition(QSlider::TicksBelow);
    offsetLayout->addWidget(m_offsetSlider);

    auto* spinLayout = new QHBoxLayout();
    m_offsetSpin = new QDoubleSpinBox(this);
    m_offsetSpin->setRange(-5.0, 5.0);
    m_offsetSpin->setDecimals(3);
    m_offsetSpin->setSingleStep(0.001);
    m_offsetSpin->setSuffix(" s");
    spinLayout->addWidget(m_offsetSpin);

    m_resetBtn = new QPushButton("Reset", this);
    m_resetBtn->setFixedWidth(70);
    spinLayout->addWidget(m_resetBtn);
    offsetLayout->addLayout(spinLayout);

    mainLayout->addWidget(offsetGroup);

    // ── Per-Source Latency ────────────────────────────────────────────────
    auto* latencyGroup = new QGroupBox("Latency Compensation", this);
    auto* latencyForm = new QFormLayout(latencyGroup);
    latencyForm->setSpacing(8);
    latencyForm->setContentsMargins(12, 16, 12, 10);

    m_uvrLatency = new QDoubleSpinBox(this);
    m_uvrLatency->setRange(-2.0, 2.0);
    m_uvrLatency->setDecimals(3);
    m_uvrLatency->setSuffix(" s");
    latencyForm->addRow("UVR Separation:", m_uvrLatency);

    m_transcriptionLatency = new QDoubleSpinBox(this);
    m_transcriptionLatency->setRange(-2.0, 2.0);
    m_transcriptionLatency->setDecimals(3);
    m_transcriptionLatency->setSuffix(" s");
    latencyForm->addRow("Transcription:", m_transcriptionLatency);

    m_playbackLatency = new QDoubleSpinBox(this);
    m_playbackLatency->setRange(-2.0, 2.0);
    m_playbackLatency->setDecimals(3);
    m_playbackLatency->setSuffix(" s");
    latencyForm->addRow("Playback:", m_playbackLatency);

    auto* noteLabel = new QLabel("Positive values = shift lyrics later.\nNegative values = shift lyrics earlier.", this);
    noteLabel->setStyleSheet("color:#888; font-size:10px; border:none;");
    noteLabel->setWordWrap(true);
    latencyForm->addRow("", noteLabel);

    mainLayout->addWidget(latencyGroup);
    mainLayout->addStretch();

    // ── Connections ──────────────────────────────────────────────────────
    connect(m_offsetSlider, &QSlider::valueChanged, this, &TimingCalibration::onSliderChanged);
    connect(m_offsetSpin, QOverload<double>::of(&QDoubleSpinBox::valueChanged),
            this, &TimingCalibration::onSpinChanged);
    connect(m_resetBtn, &QPushButton::clicked, this, &TimingCalibration::onReset);

    connect(m_uvrLatency, QOverload<double>::of(&QDoubleSpinBox::valueChanged),
            this, [this](double val) { emit latencyChanged("uvr", val); });
    connect(m_transcriptionLatency, QOverload<double>::of(&QDoubleSpinBox::valueChanged),
            this, [this](double val) { emit latencyChanged("transcription", val); });
    connect(m_playbackLatency, QOverload<double>::of(&QDoubleSpinBox::valueChanged),
            this, [this](double val) { emit latencyChanged("playback", val); });
}

void TimingCalibration::applyTheme() {
    setStyleSheet(R"(
        QWidget { background:#1e1e1e; color:#d4d4d4; }
        QGroupBox { border:1px solid #3f3f46; border-radius:6px; padding-top:16px; margin-top:6px; font-weight:bold; color:#ccc; }
        QGroupBox::title { subcontrol-origin:margin; left:10px; padding:0 4px; }
        QDoubleSpinBox { background:#2d2d30; border:1px solid #3f3f46; border-radius:4px; padding:4px 6px; color:#fff; }
        QSlider::groove:horizontal { background:#3f3f46; height:6px; border-radius:3px; }
        QSlider::handle:horizontal { background:#22d3ee; width:16px; margin:-5px 0; border-radius:8px; }
        QSlider::sub-page:horizontal { background:#22d3ee; border-radius:3px; }
        QPushButton { background:#333; border:1px solid #555; border-radius:4px; padding:5px 10px; color:#ccc; }
        QPushButton:hover { background:#404040; }
        QLabel { border:none; }
    )");
}

void TimingCalibration::setAudioClock(AudioClock* clock) {
    m_clock = clock;
    if (clock) {
        m_offsetSpin->blockSignals(true);
        m_offsetSlider->blockSignals(true);
        double offset = clock->masterOffset();
        m_offsetSpin->setValue(offset);
        m_offsetSlider->setValue(static_cast<int>(offset * 1000.0));
        m_offsetLabel->setText(QString::number(offset, 'f', 3) + " s");
        m_offsetSpin->blockSignals(false);
        m_offsetSlider->blockSignals(false);

        m_uvrLatency->setValue(clock->getLatency("uvr"));
        m_transcriptionLatency->setValue(clock->getLatency("transcription"));
        m_playbackLatency->setValue(clock->getLatency("playback"));
    }
}

void TimingCalibration::onSliderChanged(int value) {
    double seconds = value / 1000.0;
    m_offsetSpin->blockSignals(true);
    m_offsetSpin->setValue(seconds);
    m_offsetSpin->blockSignals(false);
    m_offsetLabel->setText(QString::number(seconds, 'f', 3) + " s");
    if (m_clock) m_clock->setMasterOffset(seconds);
    emit offsetChanged(seconds);
}

void TimingCalibration::onSpinChanged(double value) {
    m_offsetSlider->blockSignals(true);
    m_offsetSlider->setValue(static_cast<int>(value * 1000.0));
    m_offsetSlider->blockSignals(false);
    m_offsetLabel->setText(QString::number(value, 'f', 3) + " s");
    if (m_clock) m_clock->setMasterOffset(value);
    emit offsetChanged(value);
}

void TimingCalibration::onReset() {
    m_offsetSlider->setValue(0);
    m_uvrLatency->setValue(0);
    m_transcriptionLatency->setValue(0);
    m_playbackLatency->setValue(0);
}

} // namespace ncktv
